bigflow/configuration.py

```diff
@@ -1,44 +1,47 @@
 import os
 import io
 import pprint
+import typing
+import functools
 
 
-class EnvConfig:
-    def __init__(self,
-                 name: str,
-                 properties: dict):
-        self.name = name
-        self.properties = properties
+class EnvConfig(typing.NamedTuple):
+    name: str
+    properties: dict
 
 
 def current_env():
     """Returns current env name (specified via 'bigflow --config' option)"""
     return os.environ.get('bf_env')
 
 
+def _cache(f):
+    return functools.lru_cache(maxsize=None)(f)
+
+
 class Config:
-    def __init__(self,
-                 name: str,
-                 properties: dict,
-                 is_master: bool = True,
-                 is_default: bool = True,
-                 environment_variables_prefix: str = 'bf_'):
-        if is_master:
-            self.master_config_name = name
-
-        self.configs = {
-            name: EnvConfig(name, properties)
-        }
 
+    def __init__(self,
+        name: str,
+        properties: typing.Dict[str, str],
+        is_master: bool = True,
+        is_default: bool = True,
+    ):
+        self.master_properties = properties if is_master else {}
         self.default_env_name = None
-        self._update_default_env_name(name, is_default)
+        self.configs = {}
 
-        self.environment_variables_prefix = environment_variables_prefix
+        self._wrap_cached_methods()
+        self.add_configuration(name, properties, is_default)
 
-        self._check_docker_repository(properties)
+    def _wrap_cached_methods(self):
```

Comment:
Why do we need to cache the resolve method results?

Reply:
Just a small optimization.

Might be useful if someone heavily uses 'resolve_property' (which triggers string interpolation of the whole config; which is O(N**2)). So this code has complexity O(N**3)
  print([c.resolve_property(x) for x in c.resolve()])

Directly caching 'resolve_property' is less beneficial because it will recall 'resolve()' for each property. Complexity of the above code will be O(N**2), not O(N)

Reply:
Well, typically N should be like 5, 10, 50 maybe 100, so 100*100*100 is not a problem. But, on the other hand, who knows how a user can use this class. I leave it to @iwko and @bartoszwalacik .

Reply:
I think for this moment it is overengineering to use caching

Reply:
Some benchmarking - [10...1000] properties, 1 subst var, each property resolved once via 'resolve_property'
https://gist.github.com/anjensan/6ee025910af91e56e3825b8dd917ae5a

for 100 items the difference is quite small - just 0.28 vs 0.003 seconds
for 500 - slightly bigger - 35 vs 0.7

PS: actually I made an error. Complexity of single call is `O(N**3 * `complexity of str.replace)`

Reply:
I agree with @iwko , we don't need cache for config props

Reply:
Reverted.

Maybe add some additional comment to documentation that it is not recommended to pass 'config' into a job and invoke 'resolve_property' from it (may be in some loop), and better is to resolve all properties outside the job?

Reply:
We should solve the problem instead of informing a user how to deal with it. And because we don't want to solve the problem right now, let's also skip the recommendation.

Reply:
IMO problem should be resolved OR the fact of this problem should be tracked (create a bug?) OR we should inform user that we have that problem OR problem should be clarified as "wont-fix"

Spoke with @bartoszwalacik - agreed to slightly change `resolve()` in such way, that user will not be enforced to use 'resolve_property' & keep Config without caching. In this scenario performance of 'resolve_property' doesn't matter (as almost nobody will use it).